### ISSUE TITLE
Volumes in Kubernetes permission deny

### ISSUE THREAD

**Issue by mogody** [bitnami/bitnami-docker-etcd#8]
I deployed etcd on k8s, and the securityContext has been configured according to the documentation, but it still prompts no permission：
https://engineering.bitnami.com/articles/running-non-root-containers-on-openshift.html

this is my config:
```
spec:
        securityContext:
          runAsUser: 1001
          fsGroup: 1001
        containers:
        - image: bitnami/nginx:latest
          name: nginx
          volumeMounts:
          - mountPath: /bitnami
            name: nginx-data
```

Finally change it like this，and work fine:
```
 spec:
        initContainers:
        - name: volume-permissions
          image: busybox
          command: ['sh', '-c', 'chmod -R g+rwX /bitnami']
          volumeMounts:
          - mountPath: /bitnami
            name: nginx-data
        containers:
        - image: bitnami/nginx:latest
          name: nginx
          volumeMounts:
          - mountPath: /bitnami
            name: nginx-data
```

why use securityContext cant working?

**Comment by marcosbc** [comment]
@mogody In the docs, we recommend to use an init-container to change the permissions of the volume before mounting it:

```
      spec:
        initContainers:
        - name: volume-permissions
          image: busybox
          command: ['sh', '-c', 'chmod -R g+rwX /bitnami']
          volumeMounts:
          - mountPath: /bitnami
            name: nginx-data
        containers:
        - image: bitnami/nginx:latest
          name: nginx
          volumeMounts:
          - mountPath: /bitnami
            name: nginx-data
```

As an init-container, it will be executed before the other ones:

```
      spec:
        securityContext:
          runAsUser: 1001
          fsGroup: 1001
        containers:
        - image: bitnami/nginx:latest
          name: nginx
          volumeMounts:
          - mountPath: /bitnami
            name: nginx-data
```

Note that your Kubernetes template would need to include **both** the init-container, and the container specs. 

See the Bitnami Helm chart for etcd if you want to run etcd as a non-root user in Kubernetes: https://github.com/bitnami/charts/tree/master/bitnami/etcd

**Comment by mogody** [comment]
So what you mean is, if I want to deploy containers through none-root, do I need to configure both securityContext and initContainers?
But my test found that only initContainers can be used, but only securityContext is not enough.

In addition, how to configure subPath when deploying etcd through helm?

**Comment by marcosbc** [comment]
@mogody You should always specify securityContext when running a container as non-root. As for initContainers, It depends on your storageClass: https://kubernetes.io/docs/concepts/storage/storage-classes/

Usually with cloud defaults, it would be enough to specify securityContext with runAsUser/fsGroup, no initContainers needed. However, other storage classes such as NFS, also require you to enable the initContainer for configuring volume permissions, since they are not writable by non-root users by default.

In summary, you should configure both securityContext and initContainers.

> In addition, how to configure subPath when deploying etcd through helm?

Could you clarify what you mean? By default, you shouldn't need to modify the paths used by persistent volumes.
